go/pkg/client/tokens.go

```diff
@@ -0,0 +1,186 @@
+type tokenResp struct {
+	Lock  sync.Mutex
+	Token []byte
+	Error error
+}
+
+// getToken returns the current token, or an error if an error has occurred at some point.
+func (tk *tokenResp) getToken() ([]byte, error) {
+	tk.Lock.Lock()
+	defer tk.Lock.Unlock()
+	if tk.Error != nil {
+		return nil, tk.Error
+	} else {
+		token := make([]byte, len(tk.Token))
+		copy(token, tk.Token)
+		return token, nil
+	}
+}
+
+// setToken sets the session token to a new value.
+func (tk *tokenResp) setToken(tok []byte) {
+	tk.Lock.Lock()
+	tk.Token = tok
+	tk.Error = nil
+	tk.Lock.Unlock()
+}
+
+// setError sets an error value for the session token.
+func (tk *tokenResp) setError(err error) {
+	tk.Lock.Lock()
+	tk.Error = err
+	tk.Lock.Unlock()
+}
+
+// A tokenManager stores the current client token and sends periodic keepalive messages to refresh the client token.
+type tokenManager struct {
+	token *tokenResp // the actual client token, which gets periodically updated.
+	close func() error
+}
+
+// getToken returns the current token, or an error if an error has occurred at some point.
+func (tr *tokenManager) getToken() ([]byte, error) {
+	return tr.token.getToken()
+}
+
+// withToken attaches the current session token to a context as metadata.
+func (tr *tokenManager) withToken(ctx context.Context) (context.Context, error) {
+	tok, err := tr.getToken()
+
+	if err != nil {
+		return nil, err
+	}
+
+	return withAuthToken(ctx, tok), nil
+}
+
+func (tr *tokenManager) Close() error {
+	return tr.close()
+}
+
+// newTokenManager creates a tokenManager that begins a background goroutine that continually refreshes
+// the token so that it does not time out.
+//
+// authString is the authorization string used to get the first token.  Examples:
+//   - "Anonymous" is used for anonymous authentication.
+//   - "io.deephaven.authentication.psk.PskAuthenticationHandler <password>" is used for PSK authentication
+func newTokenManager(ctx context.Context, fs *flightStub, cfg configpb2.ConfigServiceClient, authString string) (*tokenManager, error) {
+	handshakeClient, err := fs.handshake(withAuth(ctx, authString))
+
+	if err != nil {
+		return nil, err
+	}
+
+	tkn, err := requestToken(handshakeClient, &flight.HandshakeRequest{Payload: []byte(authString)})
+
+	if err != nil {
+		return nil, err
+	}
+
+	ac, err := cfg.GetConfigurationConstants(withAuthToken(ctx, tkn), &configpb2.ConfigurationConstantsRequest{})
+
+	if err != nil {
+		return nil, err
+	}
+
+	sessionDurationStr, ok := ac.ConfigValues["http.session.durationMs"]
+
+	if !ok {
+		return nil, errors.New("server configuration constants do not contain: http.session.durationMs")
+	}
+
+	maxTimeoutMillis, err := strconv.Atoi(sessionDurationStr.GetStringValue())
+
+	if err != nil {
+		return nil, err
+	}
+
+	timeout := time.Duration(maxTimeoutMillis/2) * time.Millisecond
+
+	token := &tokenResp{Token: tkn}
+	done := make(chan bool)
+	ticker := time.NewTicker(timeout)
+
+	go func() {
+		for {
```

Comment:
In C++ we did things a little differently.  You may or may not decide they are important enough to do.

Cristian and I were quite worried about the case where a message might be lost, dropped, or massively delayed. Sending handshakes at a fixed interval, as you do here, is only reliable if they are all delivered, and delivered in a timely fashion.

To try to be more robust about this, we used the following approach instead.

1. We snoop *all* request/response traffic and we notice changes in auth token as they happen. This means when there is other traffic we don't bother sending a handshake at all, because we were able to snoop the token from some other response. We only do a handshake if there has been a `timeout`-sized period of no traffic at all.
2. The code basically has two states: normal (I have received some kind of response in the last timeout period) and handshaking (I've *not* received any kind of response in the last timeout period). When it enters the "handshaking" state, the code sends handshakes pretty aggressively (like once every 5 seconds). This allows the token refresh thing to be somewhat robust against network freezes or connection drops. But then, *any* response from the server (whether a normal response or a handshake response) tends to put the code back into the "normal" state, so in the common case you only send one handshake and then drop back to the normal state.
3. In order to account for the vagaries of round trip time and connection drops, the code keeps track of timeouts based on the local time when a (request or handshake) was *sent* rather than when a response was *received*. For example, if I sent a (request or handshake) at 12:00:00 and I receive a response at 12:00:20, I will do my timeout calculations based on when I sent the request, not when I received the response. In other words if the timeout is 5 minutes, I will consider going to handshake mode at 12:05:00, not 12:05:20.

Basically the C++ code follows this kind of pseudocode:

At any request
```
Note current local time
send the request
```

At any response to a request or a handshake
```
note the new auth header (maybe don't bother copying it if it hasn't changed)
reset the `handshake deadline` to the (time of the corresponding request) plus 5 minutes
```

The handshake thread:
```
forever:
  sleep until handshake deadline
  if current time < handshake deadline  # deadline has advanced in the meantime
     continue loop
 note current handshake request time
 send handshake message asynchronously
 set handshake deadline to now + 5 seconds
 # the above is pessimistic, because the response to this handshake message,
 # if received, will undergo standard auth header and deadline processing and
 # will advance the deadline forward ~5 minutes
```

Reply:
@jcferretti proposed (and I agree) that we should defer this work to a future PR.